src/rewards/ca/mod.rs

```diff
@@ -0,0 +1,201 @@
+        })
+        .collect();
+
+    let underbudget_funds: Funds = proposal_rewards_states
+        .values()
+        .map(|state| match state {
+            ProposalFundsState::Unfilled(value) => *value,
+            _ => Funds::from(0u64),
+        })
+        .sum();
+
+    let underbudget_rewards = underbudget_funds / Funds::from(proposal_reviews.len() as u64);
+
+    proposal_rewards_states
+        .into_iter()
+        .map(|(id, state)| {
+            let bonus_funds = approved_proposals
+                .contains(&id)
+                .then(|| bonus_proposals_rewards)
+                .unwrap_or_else(|| Funds::from(0u64));
+            let funds = match state {
+                ProposalFundsState::Unfilled(unfilled_funds) => {
+                    per_proposal_reward - unfilled_funds
+                }
+                _ => per_proposal_reward + underbudget_rewards,
+            };
+            (
+                id,
+                ProposalReward {
+                    state,
+                    funds: funds + bonus_funds,
+                },
+            )
+        })
+        .collect()
+}
+
+fn load_tickets_from_reviews(
+    proposal_reviews: &[AdvisorReviewRow],
+    rewards_slots: &ProposalRewardSlots,
+) -> TicketsDistribution {
+    let mut tickets_distribution = TicketsDistribution::new();
+    for review in proposal_reviews {
+        let entry = tickets_distribution
+            .entry(review.assessor.clone())
+            .or_insert(0);
+        let tickets_to_add = match review.score() {
+            ReviewScore::Excellent => rewards_slots.excellent_slots,
+            ReviewScore::Good => rewards_slots.good_slots,
+        };
+        *entry += tickets_to_add;
+    }
+    tickets_distribution
+}
+
+fn distribute_rewards(
+    funds: Funds,
+    cas: &TicketsDistribution,
+    rewards_slots: &ProposalRewardSlots,
+) -> CaRewards {
+    let rewards_per_ticket = funds / Funds::from(rewards_slots.filled_slots);
+    cas.iter()
+        .map(|(id, &tickets)| (id.clone(), Rewards::from(tickets) * rewards_per_ticket))
+        .collect()
+}
+
+fn lottery_rewards(
+    funds: Funds,
+    cas: &TicketsDistribution,
+    rewards_slots: &ProposalRewardSlots,
+) -> CaRewards {
+    let rewards_per_ticket = funds / Funds::from(rewards_slots.filled_slots);
+    let lottery_winnings = lottery::lottery_distribution(cas, rewards_slots.filled_slots);
+    lottery_winnings
+        .into_iter()
+        .map(|(ca, tickets_won)| (ca, Rewards::from(tickets_won) * rewards_per_ticket))
+        .collect()
+}
+
+fn calculate_ca_rewards_for_proposal(
+    proposal_reward: &ProposalReward,
+    proposal_reviews: &[AdvisorReviewRow],
+    rewards_slots: &ProposalRewardSlots,
+) -> CaRewards {
+    let ProposalReward { state, funds } = proposal_reward;
+    let tickets_distribution = load_tickets_from_reviews(proposal_reviews, rewards_slots);
+    match state {
+        ProposalFundsState::Exact | ProposalFundsState::Unfilled(_) => {
```

Comment:
actually, there's no need to have separate logic, a lottery picking `filled_slots` tickets would work for all of them

Reply:
I know about this. I made the decision so the simole implementation is straight forward, It may be a bit convoluted using a lottery just for a normal distribution.

Reply:
I think it's personal preference, but I would prefer having a single general implementation than multiple specific ones